creator-node/src/routes/files.js

```diff
@@ -240,115 +259,220 @@ const getCID = async (req, res) => {
+      RehydrateIpfsQueue.addRehydrateIpfsFromFsIfNecessaryTask(
+        CID,
+        storagePath,
+        {
+          logContext: req.logContext
+        }
+      )
+      decisionTree.push({
+        stage: `ADD_REHYDRATE_IPFS_FROM_FS`,
+        time: `${Date.now() - startMs}ms`
+      })
+    } catch (e) {
+      decisionTree.push({
+        stage: `ADD_REHYDRATE_IPFS_FROM_ERROR`,
         time: `${Date.now() - startMs}ms`,
         error: `${e.message}`
       })
+    }
+
+    /**
+     * 2. stream from FS
+     */
+    startMs = Date.now()
+    try {
+      const fsStream = await streamFromFileSystem(
+        req,
+        res,
+        storagePath,
+        false,
+        fsStats
+      )
+      decisionTree.push({
+        stage: `STREAM_FROM_FILE_SYSTEM_COMPLETE`,
+        time: `${Date.now() - startMs}ms`
+      })
+      logGetCIDDecisionTree(decisionTree, req)
+      return fsStream
+    } catch (e) {
+      decisionTree.push({
+        stage: `STREAM_FROM_FILE_SYSTEM_FAILED`,
+        time: `${Date.now() - startMs}ms`
+      })
+    }
+  }
+
+  /**
+   * If not found on FS, check if CID record is in DB, error if not found
+   */
+  startMs = Date.now()
+  try {
+    const queryResults = await models.File.findOne({
+      where: {
+        multihash: CID
+      },
+      order: [['clock', 'DESC']]
+    })
+    decisionTree.push({
+      stage: `DB_CID_QUERY`,
+      time: `${Date.now() - startMs}ms`
+    })
+
+    if (!queryResults) {
+      decisionTree.push({
+        stage: `DB_CID_QUERY_CID_NOT_FOUND`
+      })
+      logGetCIDDecisionTree(decisionTree, req)
+      return sendResponse(
+        req,
+        res,
+        errorResponseNotFound(
+          `${logPrefix} No valid file found for provided CID`
+        )
+      )
+    } else if (queryResults.type === 'dir') {
+      decisionTree.push({
+        stage: `DB_CID_QUERY_CONFIRMED_DIR`
+      })
       logGetCIDDecisionTree(decisionTree, req)
       return sendResponse(
         req,
         res,
-        errorResponseServerError(`${logPrefix} DB query failed`)
+        errorResponseBadRequest('this dag node is a directory')
       )
+    } else {
+      decisionTree.push({
+        stage: `DB_CID_QUERY_CID_FOUND`
+      })
+      // This should never happen, logging in case it does
+      if (storagePath !== queryResults.storagePath) {
+        req.logger.error(
+          `${logPrefix} Found unexpected storagePath. Expected ${storagePath}, found ${queryResults.storagePath}.`
+        )
+      }
+      storagePath = queryResults.storagePath
     }
+  } catch (e) {
+    decisionTree.push({
+      stage: `DB_CID_QUERY_ERROR`,
+      time: `${Date.now() - startMs}ms`,
+      error: `${e.message}`
+    })
+    logGetCIDDecisionTree(decisionTree, req)
+    return sendResponse(
+      req,
+      res,
+      errorResponseServerError(`${logPrefix} DB query failed`)
+    )
   }
 
   /**
-   * If found in DB, but not file system:
+   * If found in DB (means not found in FS):
```

Comment:
this comment was so helpful. at first i was like if not found in db why do network lookup...?

Reply:
yeah lol